### PR TITLE
Fetch submodules if supported, and warn if submodules are used but not supported

### PR DESCRIPTION
**Pull request by tfc**
Referencing @awidegreen's PR https://github.com/nmattia/niv/pull/351, this one here does the same but also adds a warning if submodules are used but not supported by the user's nix version.

In that corner case, the user get's at least a warning trace on their shell that explains what's wrong here. Builds can still fail, but not silently without some hint on the reason.

### CODE REVIEW

File: nix/sources.nix
Comment (nmattia) [inline nix/sources.nix:35]: nit: how about moving `nixSupportsSubmodules` to the `let` on line 37? I don't think it's used outside of it

Comment (tfc) [reply nix/sources.nix]: You're right. I applied and pushed your suggestion.

**Review by nmattia — APPROVED** [review]
LGTM, thanks!

**Comment by tfc** [comment]
@nmattia i am not sure where the error in the CI comes from. Looking at the source code we have a `$(embedFile "nix/sources.nix")` which should automatically be the right thing, right?

I currently don't see how i would fix that, can you give me a pointer?

**Comment by tfc** [comment]
Interesting, thank you. I added it all. Let's see what the pipeline says.

**Comment by SuperSandro2000** [comment]
> if submodules are used but not supported by the user's nix version.

I don't think it should warn there. Often submodules are vendored projects which can be replaced by system libraries.

**Comment by tfc** [comment]
> I don't think it should warn there. Often submodules are vendored projects which can be replaced by system libraries.

I am sure such repos exist. In such cases you are still free to set `submodules = false` and use it as you describe.

If repos *cannot* live without their submodules, then we get silent failures here, which is very bad.

**Comment by tfc** [comment]
@nmattia it seems this time everything looks green! Thank you for your help.

**Comment by tfc** [comment]
Is anything left to do before this can be merged?

File: nix/sources.nix
Comment (nmattia) [inline nix/sources.nix:180]: Any idea why this got (re-)formatted?

**Review by nmattia — COMMENTED** [review]
The weird formatting thing aside, LGTM!

**Comment by tfc** [comment]
@nmattia ok, does the weird formatting block the pull request now, so do i need to analyze why the tool decided to format this unrelated block of code, or is it ok?

**Comment by tfc** [comment]
@nmattia i am happy to do anything that's necessary to make this mergeable. Right now i don't know what's missing and i also don't know if there is anything else that you need (time to review because you are busy right now (in that case shall we add any other reviewers? do we need more testing? or anything else having to be merged before this or something?), so please let me know.

**Comment by nmattia** [comment]
@tfc my bad, I was out for a few days! Should have been clearer: I'm confused as to why the formatting was changed around line 170, but it looks good to me nonetheless! For me you can merge at will :) You _could_ revert the formatting on line 170 (maybe it was done by your editor? either way nixpkgs-fmt doesn't seem to mind, but probably cleaner to remove it so that future `git blame`rs don't get confused) but you could also just merge as is. Your call, and thanks a lot for the contribution!

**Comment by tfc** [comment]
Ok. So the additional format change was my mistake, sorry for the resulting noise. I removed it.

Btw. once the tests are confirmed green: i have no merge rights so i can't follow the "you could also just merge"

**Comment by nmattia** [comment]
There you go, thanks a lot!

**Pull request merged**
